eventlet/hubs/hub.py

```diff
@@ -224,15 +224,23 @@ def remove(self, listener):
 
         fileno = listener.fileno
         evtype = listener.evtype
-        self.listeners[evtype].pop(fileno, None)
-        # migrate a secondary listener to be the primary listener
-        if fileno in self.secondaries[evtype]:
-            sec = self.secondaries[evtype].get(fileno, None)
-            if not sec:
-                return
-            self.listeners[evtype][fileno] = sec.pop(0)
-            if not sec:
-                del self.secondaries[evtype][fileno]
+        if listener is self.listeners[evtype].get(fileno):
+            self.listeners[evtype].pop(fileno, None)
+            # migrate a secondary listener to be the primary listener
+            if fileno in self.secondaries[evtype]:
+                sec = self.secondaries[evtype].get(fileno, None)
+                if not sec:
+                    return
+                self.listeners[evtype][fileno] = sec.pop(0)
+                if not sec:
+                    del self.secondaries[evtype][fileno]
+        else:
+            sec = [l for l in self.secondaries[evtype].get(fileno, [])
+                   if l is not listener]
+            if sec:
+                self.secondaries[evtype][fileno] = sec
+            else:
+                self.secondaries[evtype].pop(fileno, None)
```

Comment:
This stuff is the key fix -- if we're not the primary listener, *don't touch `self.listeners`*! Love it!

Reply:
@tipabu hey Tim do you think #645 that was merged is in good shape regarding this comment? Thanks.

Reply:
Yeah, absolutely -- it's even cleaner, down to just
```
            self.secondaries[evtype][fileno].remove(listener)
            if not self.secondaries[evtype][fileno]:
                del self.secondaries[evtype][fileno]
```
Clay even addressed my other thoughts while he was making that `remove(listener)` still pass tests. Kudos!

Sorry for the extra noise after #645 was approved; I'd already started looking at this and didn't notice everything was already taken care of before I left the review.

Reply:
It's not noise, but a signal (yes) that there are multiple people so well familiar with hub code. That is awesome.